File: src/normalize.ts
import { CycleCloudRequestError } from "./errors.js";

export interface ClusterSummary {
    readonly name: string;
    readonly state?: string;
    readonly targetState?: string;
    readonly fixedNodeDefinitions: number;
    readonly nodeArrayCount: number;
    readonly arrayNodeCount: number;
    readonly configuredNodeCount: number;
}

export interface ClusterListResult {
    readonly clusters: readonly ClusterSummary[];
    readonly total: number;
    readonly returned: number;
    readonly truncated: boolean;
}

export interface NodeArraySummary {
    readonly template: string;
    readonly state?: string;
    readonly targetState?: string;
    readonly count: number;
    readonly coreCount?: number;
}

export interface FixedNodeSummary {
    readonly id?: string;
    readonly name: string;
    readonly template?: string;
    readonly state?: string;
    readonly targetState?: string;
}

export interface ClusterDetailResult {
    readonly cluster: {
        readonly name: string;
        readonly state?: string;
        readonly targetState?: string;
        readonly nodeArrays: readonly NodeArraySummary[];
        readonly nodeArrayTotal: number;
        readonly nodeArrayReturned: number;
        readonly nodeArraysTruncated: boolean;
        readonly fixedNodes: readonly FixedNodeSummary[];
        readonly fixedNodeDefinitionsTotal: number;
        readonly fixedNodeDefinitionsReturned: number;
        readonly fixedNodeDefinitionsTruncated: boolean;
        readonly arrayNodeCount: number;
        readonly configuredNodeCount: number;
    };
}

export interface BucketStatus {
    readonly bucketId: string;
    readonly machineType?: string;
    readonly valid: boolean;
    readonly invalidReason?: string;
    readonly maxCount: number;
    readonly maxCoreCount: number;
    readonly quotaCount: number;
    readonly quotaCoreCount: number;
    readonly consumedCoreCount: number;
    readonly activeCount: number;
    readonly activeCoreCount: number;
    readonly availableCount: number;
    readonly availableCoreCount: number;
    readonly lastCapacityFailure?: number;
    readonly spotPlacementScore?: string;
}

export interface NodeArrayStatus {
    readonly name: string;
    readonly maxCount: number;
    readonly maxCoreCount: number;
    readonly buckets: readonly BucketStatus[];
    readonly bucketTotal: number;
    readonly bucketReturned: number;
    readonly bucketsTruncated: boolean;
}

export interface ClusterStatusResult {
    readonly status: {
        readonly clusterName: string;
        readonly state?: string;
        readonly targetState?: string;
        readonly maxCount: number;
        readonly maxCoreCount: number;
        readonly nodeArrays: readonly NodeArrayStatus[];
        readonly nodeArrayTotal: number;
        readonly nodeArrayReturned: number;
        readonly nodeArraysTruncated: boolean;
        readonly bucketTotal: number;
        readonly bucketReturned: number;
        readonly bucketsTruncated: boolean;
    };
}

export interface ClusterIssue {
    readonly name: string;
    readonly severity: "Error" | "Warning";
    readonly nodeCount: number;
    readonly message?: string;
    readonly detail?: string;
    readonly recommendation?: string;
    readonly textTruncated: boolean;
}

export interface ClusterIssues {
    readonly available: true;
    readonly items: readonly ClusterIssue[];
    readonly total: number;
    readonly returned: number;
    readonly truncated: boolean;
}

export type ClusterStatusWithIssues = {
    readonly status: ClusterStatusResult["status"] & {
        readonly issues:
            | ClusterIssues
            | {
                  readonly available: false;
                  readonly warning: string;
              };
    };
};

type ConsumedFields = ReadonlyMap<string, unknown>;

export function normalizeClusterIssues(
    raw: unknown,
    limit: number,
): ClusterIssues {
    validateLimit(limit, 0, 100);
    if (!Array.isArray(raw)) invalidResponse();
    const issues: ClusterIssue[] = [];
    for (const value of raw) {
        const fields = consumedFields(value, [
            "name",
            "status",
            "nodecount",
            "message",
            "detail",
            "recommendation",
        ]);
        const severity = requiredString(fields, "status", 128);
        if (severity === "OK" || severity === "Pending") continue;
        if (severity !== "Error" && severity !== "Warning") invalidResponse();
        let textTruncated = false;
        const text: Record<string, string> = {};
        for (const key of ["message", "detail", "recommendation"]) {
            const value = fields.get(key);
            if (value === undefined || value === null) continue;
            if (typeof value !== "string" || !isWellFormed(value))
                invalidResponse();
            const characters = [
                ...value.replace(/[\u0000-\u001f\u007f-\u009f]/gu, " "),
            ];
            textTruncated ||= characters.length > 2048;
            text[key] =
                characters.length > 2048
                    ? characters.slice(0, 2047).join("") + "…"
                    : characters.join("");
        }
        issues.push({
            name: requiredString(fields, "name", 256),
            severity,
            nodeCount: requiredNonNegativeInteger(
                requiredField(fields, "nodecount"),
            ),
            ...text,
            textTruncated,
        });
    }
    issues.sort(
        (left, right) =>
            compareNames(left.severity, right.severity) ||
            compareNames(left.name, right.name) ||
            compareNames(left.message ?? "", right.message ?? ""),
    );
    const items = issues.slice(0, limit);
    return {
        available: true,
        items,
        total: issues.length,
        returned: items.length,
        truncated: items.length < issues.length,
    };
}

export function normalizeClusterList(
    raw: unknown,
    limit: number,
): ClusterListResult {
    validateLimit(limit, 1, 200);
    if (!Array.isArray(raw)) invalidResponse();

    const clusters = raw
        .map(normalizeClusterSummary)
        .sort((left, right) => compareNames(left.name, right.name));
    const returned = clusters.slice(0, limit);
    return {
        clusters: returned,
        total: clusters.length,
        returned: returned.length,
        truncated: returned.length < clusters.length,
    };
}

export function normalizeCluster(
    raw: unknown,
    requestedName: string,
    fixedNodeLimit: number,
    nodeArrayLimit: number,
): ClusterDetailResult {
    validateLimit(fixedNodeLimit, 0, 200);
    validateLimit(nodeArrayLimit, 0, 100);
    if (!Array.isArray(raw)) invalidResponse();
    if (raw.length === 0)
        throw new CycleCloudRequestError("cluster_not_found", false);
    if (raw.length !== 1) invalidResponse();

    const fields = consumedFields(raw[0], [
        "clustername",
        "state",
        "targetstate",
        "nodes",
        "nodearrays",
    ]);
    const name = requiredString(fields, "clustername", 256);
    if (name !== requestedName) invalidResponse();

    const state = optionalString(fields, "state", 128);
    const targetState = optionalString(fields, "targetstate", 128);
    const fixedNodes = optionalArray(fields, "nodes")
        .map(normalizeFixedNode)
        .sort((left, right) => compareNames(left.name, right.name));
    const nodeArrays = optionalArray(fields, "nodearrays")
        .map(normalizeNodeArray)
        .sort((left, right) => compareNames(left.template, right.template));
    const arrayNodeCount = checkedSum(
        nodeArrays.map((nodeArray) => nodeArray.count),
    );
    const configuredNodeCount = checkedAdd(fixedNodes.length, arrayNodeCount);
    const returnedFixedNodes = fixedNodes.slice(0, fixedNodeLimit);
    const returnedNodeArrays = nodeArrays.slice(0, nodeArrayLimit);

    return {
        cluster: {
            name,
            ...optionalProperty("state", state),
            ...optionalProperty("targetState", targetState),
            nodeArrays: returnedNodeArrays,
            nodeArrayTotal: nodeArrays.length,
            nodeArrayReturned: returnedNodeArrays.length,
            nodeArraysTruncated: returnedNodeArrays.length < nodeArrays.length,
            fixedNodes: returnedFixedNodes,
            fixedNodeDefinitionsTotal: fixedNodes.length,
            fixedNodeDefinitionsReturned: returnedFixedNodes.length,
            fixedNodeDefinitionsTruncated:
                returnedFixedNodes.length < fixedNodes.length,
            arrayNodeCount,
            configuredNodeCount,
        },
    };
}

export function normalizeClusterStatus(
    raw: unknown,
    clusterName: string,
    nodeArrayLimit: number,
    bucketLimit: number,
): ClusterStatusResult {
    validateLimit(nodeArrayLimit, 0, 50);
    validateLimit(bucketLimit, 0, 50);
    const record = requiredRecord(raw);
    const state = optionalWireString(record.state, 128);
    const targetState = optionalWireString(record.targetState, 128);
    const maxCount = requiredNonNegativeInteger(record.maxCount);
    const maxCoreCount = requiredNonNegativeInteger(record.maxCoreCount);
    if (!Array.isArray(record.nodearrays)) invalidResponse();

    const nodeArrays = record.nodearrays
        .map(normalizeNodeArrayStatus)
        .sort((left, right) => compareNames(left.name, right.name));
    const bucketTotal = checkedSum(
        nodeArrays.map((nodeArray) => nodeArray.buckets.length),
    );
    const selectedNodeArrays = nodeArrays.slice(0, nodeArrayLimit);
    let remainingBucketBudget = 500;
    const returnedNodeArrays = selectedNodeArrays.map(
        (nodeArray): NodeArrayStatus => {
            const count = Math.min(
                nodeArray.buckets.length,
                bucketLimit,
                remainingBucketBudget,
            );
            const buckets = nodeArray.buckets.slice(0, count);
            remainingBucketBudget -= buckets.length;
            return {
                name: nodeArray.name,
                maxCount: nodeArray.maxCount,
                maxCoreCount: nodeArray.maxCoreCount,
                buckets,
                bucketTotal: nodeArray.buckets.length,
                bucketReturned: buckets.length,
                bucketsTruncated: buckets.length < nodeArray.buckets.length,
            };
        },
    );
    const bucketReturned = checkedSum(
        returnedNodeArrays.map((nodeArray) => nodeArray.bucketReturned),
    );

    return {
        status: {
            clusterName,
            ...optionalProperty("state", state),
            ...optionalProperty("targetState", targetState),
            maxCount,
            maxCoreCount,
            nodeArrays: returnedNodeArrays,
            nodeArrayTotal: nodeArrays.length,
            nodeArrayReturned: returnedNodeArrays.length,
            nodeArraysTruncated: returnedNodeArrays.length < nodeArrays.length,
            bucketTotal,
            bucketReturned,
            bucketsTruncated: bucketReturned < bucketTotal,
        },
    };
}

function normalizeClusterSummary(value: unknown): ClusterSummary {
    const fields = consumedFields(value, [
        "clustername",
        "state",
        "targetstate",
        "nodes",
        "nodearrays",
    ]);
    const name = requiredString(fields, "clustername", 256);
    const state = optionalString(fields, "state", 128);
    const targetState = optionalString(fields, "targetstate", 128);
    const fixedNodes = optionalArray(fields, "nodes");
    const nodeArrays = optionalArray(fields, "nodearrays");
    const arrayNodeCount = checkedSum(
        nodeArrays.map((nodeArray) =>
            requiredNonNegativeInteger(
                requiredField(consumedFields(nodeArray, ["count"]), "count"),
            ),
        ),
    );
    return {
        name,
        ...optionalProperty("state", state),
        ...optionalProperty("targetState", targetState),
        fixedNodeDefinitions: fixedNodes.length,
        nodeArrayCount: nodeArrays.length,
        arrayNodeCount,
        configuredNodeCount: checkedAdd(fixedNodes.length, arrayNodeCount),
    };
}

function normalizeFixedNode(value: unknown): FixedNodeSummary {
    const fields = consumedFields(value, [
        "nodeid",
        "name",
        "template",
        "state",
        "targetstate",
    ]);
    const id = optionalString(fields, "nodeid", 256);
    const name = requiredString(fields, "name", 256);
    const template = optionalString(fields, "template", 256);
    const state = optionalString(fields, "state", 128);
    const targetState = optionalString(fields, "targetstate", 128);
    return {
        ...optionalProperty("id", id),
        name,
        ...optionalProperty("template", template),
        ...optionalProperty("state", state),
        ...optionalProperty("targetState", targetState),
    };
}

function normalizeNodeArray(value: unknown): NodeArraySummary {
    const fields = consumedFields(value, [
        "template",
        "state",
        "targetstate",
        "count",
        "corecount",
    ]);
    const template = requiredString(fields, "template", 256);
    const state = optionalString(fields, "state", 128);
    const targetState = optionalString(fields, "targetstate", 128);
    const count = requiredNonNegativeInteger(requiredField(fields, "count"));
    const coreCountValue = fields.get("corecount");
    const coreCount =
        coreCountValue === undefined
            ? undefined
            : requiredNonNegativeInteger(coreCountValue);
    return {
        template,
        ...optionalProperty("state", state),
        ...optionalProperty("targetState", targetState),
        count,
        ...optionalProperty("coreCount", coreCount),
    };
}

function normalizeNodeArrayStatus(
    value: unknown,
): Omit<
    NodeArrayStatus,
    "bucketTotal" | "bucketReturned" | "bucketsTruncated"
> {
    const record = requiredRecord(value);
    if (!Array.isArray(record.buckets)) invalidResponse();
    return {
        name: requiredWireString(record.name, 256),
        maxCount: requiredNonNegativeInteger(record.maxCount),
        maxCoreCount: requiredNonNegativeInteger(record.maxCoreCount),
        buckets: record.buckets
            .map(normalizeBucketStatus)
            .sort((left, right) => compareNames(left.bucketId, right.bucketId)),
    };
}

function normalizeBucketStatus(value: unknown): BucketStatus {
    const record = requiredRecord(value);
    const definition =
        record.definition === undefined
            ? undefined
            : requiredRecord(record.definition);
    const machineType =
        definition === undefined
            ? undefined
            : optionalWireString(definition.machineType, 256);
    const invalidReason = optionalWireString(record.invalidReason, 256);
    const lastCapacityFailure = optionalFiniteNumber(
        record.lastCapacityFailure,
    );
    const spotPlacementScore = optionalWireString(
        record.spotPlacementScore,
        256,
    );
    if (typeof record.valid !== "boolean") invalidResponse();
    return {
        bucketId: requiredWireString(record.bucketId, 256),
        ...optionalProperty("machineType", machineType),
        valid: record.valid,
        ...optionalProperty("invalidReason", invalidReason),
        maxCount: requiredNonNegativeInteger(record.maxCount),
        maxCoreCount: requiredNonNegativeInteger(record.maxCoreCount),
        quotaCount: requiredNonNegativeInteger(record.quotaCount),
        quotaCoreCount: requiredNonNegativeInteger(record.quotaCoreCount),
        consumedCoreCount: requiredNonNegativeInteger(record.consumedCoreCount),
        activeCount: requiredNonNegativeInteger(record.activeCount),
        activeCoreCount: requiredNonNegativeInteger(record.activeCoreCount),
        availableCount: requiredNonNegativeInteger(record.availableCount),
        availableCoreCount: requiredNonNegativeInteger(
            record.availableCoreCount,
        ),
        ...optionalProperty("lastCapacityFailure", lastCapacityFailure),
        ...optionalProperty("spotPlacementScore", spotPlacementScore),
    };
}

export function consumedFields(
    value: unknown,
    consumedNames: readonly string[],
): ConsumedFields {
    const record = requiredRecord(value);
    const allowed = new Set(consumedNames);
    const result = new Map<string, unknown>();
    for (const [key, fieldValue] of Object.entries(record)) {
        const normalizedKey = asciiLowercase(key);
        if (!allowed.has(normalizedKey)) continue;
        if (result.has(normalizedKey)) invalidResponse();
        result.set(normalizedKey, fieldValue);
    }
    return result;
}

export function requiredRecord(value: unknown): Record<string, unknown> {
    if (!isRecord(value)) invalidResponse();
    return value;
}

function isRecord(value: unknown): value is Record<string, unknown> {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}

function requiredField(fields: ConsumedFields, name: string): unknown {
    if (!fields.has(name)) invalidResponse();
    return fields.get(name);
}

function optionalArray(
    fields: ConsumedFields,
    name: string,
): readonly unknown[] {
    const value = fields.get(name);
    if (value === undefined) return [];
    if (!Array.isArray(value)) invalidResponse();
    return value;
}

function requiredString(
    fields: ConsumedFields,
    name: string,
    maximumScalars: number,
): string {
    return requiredWireString(requiredField(fields, name), maximumScalars);
}

function optionalString(
    fields: ConsumedFields,
    name: string,
    maximumScalars: number,
): string | undefined {
    return optionalWireString(fields.get(name), maximumScalars);
}

export function requiredWireString(
    value: unknown,
    maximumScalars: number,
): string {
    const result = optionalWireString(value, maximumScalars);
    if (result === undefined || result.length === 0) invalidResponse();
    return result;
}

function optionalWireString(
    value: unknown,
    maximumScalars: number,
): string | undefined {
    if (value === undefined) return undefined;
    if (
        typeof value !== "string" ||
        !isWellFormed(value) ||
        /[\u0000-\u001f\u007f-\u009f]/u.test(value)
    ) {
        invalidResponse();
    }
    if ([...value].length > maximumScalars) invalidResponse();
    return value;
}

function requiredNonNegativeInteger(value: unknown): number {
    if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 0)
        invalidResponse();
    return value;
}

function optionalFiniteNumber(value: unknown): number | undefined {
    if (value === undefined) return undefined;
    if (typeof value !== "number" || !Number.isFinite(value)) invalidResponse();
    return value;
}

function checkedSum(values: readonly number[]): number {
    return values.reduce(checkedAdd, 0);
}

function checkedAdd(left: number, right: number): number {
    const result = left + right;
    if (!Number.isSafeInteger(result) || result < 0) invalidResponse();
    return result;
}

function validateLimit(value: number, minimum: number, maximum: number): void {
    if (!Number.isSafeInteger(value) || value < minimum || value > maximum)
        invalidResponse();
}

export function compareNames(left: string, right: string): number {
    const foldedLeft = asciiLowercase(left);
    const foldedRight = asciiLowercase(right);
    if (foldedLeft < foldedRight) return -1;
    if (foldedLeft > foldedRight) return 1;
    if (left < right) return -1;
    if (left > right) return 1;
    return 0;
}

function asciiLowercase(value: string): string {
    return value.replace(/[A-Z]/g, (character) =>
        String.fromCharCode(character.charCodeAt(0) + 32),
    );
}

function isWellFormed(value: string): boolean {
    for (let index = 0; index < value.length; index += 1) {
        const code = value.charCodeAt(index);
        if (code >= 0xd800 && code <= 0xdbff) {
            const next = value.charCodeAt(index + 1);
            if (!(next >= 0xdc00 && next <= 0xdfff)) return false;
            index += 1;
        } else if (code >= 0xdc00 && code <= 0xdfff) {
            return false;
        }
    }
    return true;
}

function optionalProperty<Value>(
    key: string,
    value: Value | undefined,
): Record<string, Value> {
    return value === undefined ? {} : { [key]: value };
}

function invalidResponse(): never {
    throw new CycleCloudRequestError("invalid_response", false);
}

File: python/cyclecloud_agent_inspect/normalize.py
"""Closed-allowlist normalization of CycleCloud query and public status data."""
import math
import re

from .contract import (
    MISSING, ascii_lowercase, name_sort_key, omit_missing, optional_wire_string,
    required_record, required_rows, required_wire_string, safe_integer,
    validate_limit, well_formed_text,
)
from .errors import cluster_not_found, invalid_response


def consumed_fields(value, consumed_names):
    allowed = set(consumed_names)
    result = {}
    for key, field_value in required_record(value).items():
        normalized = ascii_lowercase(key)
        if normalized not in allowed:
            continue
        if normalized in result:
            invalid_response()
        result[normalized] = field_value
    return result


def _optional_array(fields, name):
    return required_rows(fields[name]) if name in fields else []


def _optional_strings(fields, definitions):
    return omit_missing({output: optional_wire_string(fields.get(source, MISSING), limit)
                         for output, source, limit in definitions})


def _checked_sum(values):
    return safe_integer(sum(values), 0)


def _cluster_fields(value):
    return consumed_fields(value, ("clustername", "state", "targetstate", "nodes", "nodearrays"))


def _states(fields):
    return _optional_strings(fields, (("state", "state", 128), ("targetState", "targetstate", 128)))


def _cluster_summary(value):
    fields = _cluster_fields(value)
    name = required_wire_string(fields.get("clustername", MISSING), 256)
    states = _states(fields)
    fixed_nodes = _optional_array(fields, "nodes")
    node_arrays = _optional_array(fields, "nodearrays")
    array_count = _checked_sum([
        safe_integer(consumed_fields(node, ["count"]).get("count", MISSING), 0)
        for node in node_arrays
    ])
    return {"name": name, **states, "fixedNodeDefinitions": len(fixed_nodes),
            "nodeArrayCount": len(node_arrays), "arrayNodeCount": array_count,
            "configuredNodeCount": _checked_sum([len(fixed_nodes), array_count])}


def normalize_cluster_list(raw, limit=50):
    limit = validate_limit(limit, 1, 200)
    clusters = sorted((_cluster_summary(row) for row in required_rows(raw)), key=lambda row: name_sort_key(row["name"]))
    selected = clusters[:limit]
    return {"clusters": selected, "total": len(clusters), "returned": len(selected), "truncated": len(selected) < len(clusters)}


def _fixed_node(value):
    fields = consumed_fields(value, ("nodeid", "name", "template", "state", "targetstate"))
    return {"name": required_wire_string(fields.get("name", MISSING), 256),
            **_optional_strings(fields, (("id", "nodeid", 256), ("template", "template", 256))), **_states(fields)}


def _node_array(value):
    fields = consumed_fields(value, ("template", "state", "targetstate", "count", "corecount"))
    result = {"template": required_wire_string(fields.get("template", MISSING), 256),
              **_states(fields), "count": safe_integer(fields.get("count", MISSING), 0)}
    if "corecount" in fields:
        result["coreCount"] = safe_integer(fields["corecount"], 0)
    return result


def normalize_cluster(raw, cluster_name, fixed_node_limit=50, node_array_limit=50):
    fixed_node_limit = validate_limit(fixed_node_limit, 0, 200)
    node_array_limit = validate_limit(node_array_limit, 0, 100)
    rows = required_rows(raw)
    if not rows:
        cluster_not_found()
    if len(rows) != 1:
        invalid_response()
    fields = _cluster_fields(rows[0])
    name = required_wire_string(fields.get("clustername", MISSING), 256)
    if name != cluster_name:
        invalid_response()
    states = _states(fields)
    fixed_nodes = sorted((_fixed_node(row) for row in _optional_array(fields, "nodes")), key=lambda row: name_sort_key(row["name"]))
    node_arrays = sorted((_node_array(row) for row in _optional_array(fields, "nodearrays")), key=lambda row: name_sort_key(row["template"]))
    array_count = _checked_sum([node["count"] for node in node_arrays])
    selected_fixed = fixed_nodes[:fixed_node_limit]
    selected_arrays = node_arrays[:node_array_limit]
    return {"cluster": {"name": name, **states,
                        "nodeArrays": selected_arrays, "nodeArrayTotal": len(node_arrays), "nodeArrayReturned": len(selected_arrays),
                        "nodeArraysTruncated": len(selected_arrays) < len(node_arrays),
                        "fixedNodes": selected_fixed, "fixedNodeDefinitionsTotal": len(fixed_nodes),
                        "fixedNodeDefinitionsReturned": len(selected_fixed),
                        "fixedNodeDefinitionsTruncated": len(selected_fixed) < len(fixed_nodes),
                        "arrayNodeCount": array_count, "configuredNodeCount": _checked_sum([len(fixed_nodes), array_count])}}


def _bucket_status(value):
    record = required_record(value)
    result = {"bucketId": required_wire_string(record.get("bucketId", MISSING), 256)}
    if "definition" in record:
        definition = required_record(record["definition"])
        machine = optional_wire_string(definition.get("machineType", MISSING), 256)
        if machine is not MISSING:
            result["machineType"] = machine
    if not isinstance(record.get("valid"), bool):
        invalid_response()
    result["valid"] = record["valid"]
    for field in ("maxCount", "maxCoreCount", "quotaCount", "quotaCoreCount", "consumedCoreCount",
                  "activeCount", "activeCoreCount", "availableCount", "availableCoreCount"):
        result[field] = safe_integer(record.get(field, MISSING), 0)
    result.update(_optional_strings(record, (("invalidReason", "invalidReason", 256), ("spotPlacementScore", "spotPlacementScore", 256))))
    if "lastCapacityFailure" in record:
        failure = record["lastCapacityFailure"]
        if isinstance(failure, bool) or not isinstance(failure, (int, float)):
            invalid_response()
        try:
            if not math.isfinite(failure):
                invalid_response()
        except OverflowError:
            invalid_response()
        result["lastCapacityFailure"] = failure
    return result


def _node_array_status(value):
    record = required_record(value)
    return {"name": required_wire_string(record.get("name", MISSING), 256),
            "maxCount": safe_integer(record.get("maxCount", MISSING), 0),
            "maxCoreCount": safe_integer(record.get("maxCoreCount", MISSING), 0),
            "buckets": sorted((_bucket_status(row) for row in required_rows(record.get("buckets"))),
                              key=lambda row: name_sort_key(row["bucketId"]))}


def normalize_cluster_status(raw, cluster_name, node_array_limit=20, bucket_limit=20):
    node_array_limit = validate_limit(node_array_limit, 0, 50)
    bucket_limit = validate_limit(bucket_limit, 0, 50)
    record = required_record(raw)
    states = _optional_strings(record, (("state", "state", 128), ("targetState", "targetState", 128)))
    max_count = safe_integer(record.get("maxCount", MISSING), 0)
    max_core_count = safe_integer(record.get("maxCoreCount", MISSING), 0)
    arrays = sorted((_node_array_status(row) for row in required_rows(record.get("nodearrays"))), key=lambda row: name_sort_key(row["name"]))
    bucket_total = _checked_sum([len(row["buckets"]) for row in arrays])
    selected = []
    remaining = 500
    for row in arrays[:node_array_limit]:
        count = min(len(row["buckets"]), bucket_limit, remaining)
        buckets = row["buckets"][:count]
        remaining -= count
        selected.append({**row, "buckets": buckets, "bucketTotal": len(row["buckets"]),
                         "bucketReturned": count, "bucketsTruncated": count < len(row["buckets"])})
    returned = _checked_sum([row["bucketReturned"] for row in selected])
    return {"status": {"clusterName": cluster_name, **states, "maxCount": max_count, "maxCoreCount": max_core_count,
                       "nodeArrays": selected, "nodeArrayTotal": len(arrays), "nodeArrayReturned": len(selected),
                       "nodeArraysTruncated": len(selected) < len(arrays), "bucketTotal": bucket_total,
                       "bucketReturned": returned, "bucketsTruncated": returned < bucket_total}}


def normalize_cluster_issues(raw, issue_limit=20):
    issue_limit = validate_limit(issue_limit, 0, 100)
    issues = []
    for row in required_rows(raw):
        fields = consumed_fields(row, ("name", "status", "nodecount", "message", "detail", "recommendation"))
        severity = required_wire_string(fields.get("status", MISSING), 128)
        if severity in ("OK", "Pending"):
            continue
        if severity not in ("Error", "Warning"):
            invalid_response()
        text = {}
        truncated = False
        for field in ("message", "detail", "recommendation"):
            value = fields.get(field)
            if value is None:
                continue
            value = re.sub(r"[\x00-\x1f\x7f-\x9f]", " ", well_formed_text(value))
            truncated = truncated or len(value) > 2048
            text[field] = value[:2047] + "…" if len(value) > 2048 else value
        issues.append({"name": required_wire_string(fields.get("name", MISSING), 256), "severity": severity,
                       "nodeCount": safe_integer(fields.get("nodecount", MISSING), 0), **text, "textTruncated": truncated})
    issues.sort(key=lambda row: (name_sort_key(row["severity"]), name_sort_key(row["name"]), name_sort_key(row.get("message", ""))))
    selected = issues[:issue_limit]
    return {"available": True, "items": selected, "total": len(issues), "returned": len(selected), "truncated": len(selected) < len(issues)}
